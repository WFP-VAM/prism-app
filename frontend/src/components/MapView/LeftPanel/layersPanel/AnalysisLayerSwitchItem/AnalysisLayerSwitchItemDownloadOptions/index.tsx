import React, { memo, useMemo, useCallback, useState } from 'react';
import { IconButton, Menu, MenuItem, Tooltip } from '@material-ui/core';
import GetAppIcon from '@material-ui/icons/GetApp';
import { useSafeTranslation } from '../../../../../../i18n';
import { downloadToFile } from '../../../../utils';
import {
  BaselineLayerResult,
  downloadCSVFromTableData,
  ExposedPopulationResult,
  generateAnalysisFilename,
  PolygonAnalysisResult,
  useAnalysisTableColumns,
} from '../../../../../../utils/analysis-utils';

const AnalysisLayerSwitchItemDownloadOptions = memo(
  ({
    selected,
    analysisData,
    analysisResultSortByKey,
    analysisResultSortOrder,
  }: AnalysisLayerSwitchItemDownloadOptionsProps) => {
    const [
      downloadMenuAnchorEl,
      setDownloadMenuAnchorEl,
    ] = useState<HTMLElement | null>(null);

    const { translatedColumns } = useAnalysisTableColumns(analysisData);

    const { t } = useSafeTranslation();

    const featureCollection = useMemo(() => {
      return analysisData?.featureCollection;
    }, [analysisData]);

    const doesLayerAcceptCSVDownload = useMemo(() => {
      return (
        analysisData &&
        (analysisData instanceof BaselineLayerResult ||
          analysisData instanceof PolygonAnalysisResult)
      );
    }, [analysisData]);

    const handleDownloadMenuClose = useCallback(() => {
      setDownloadMenuAnchorEl(null);
    }, []);

    const handleDownloadMenuOpen = useCallback(
      (event: React.MouseEvent<HTMLElement>) => {
        setDownloadMenuAnchorEl(event.currentTarget);
      },
      [],
    );

    const renderedDownloadButton = useMemo(() => {
      if (!selected) {
        return (
          <IconButton disabled={!selected} onClick={handleDownloadMenuOpen}>
            <GetAppIcon />
          </IconButton>
        );
      }
      return (
        <Tooltip title="Download">
          <IconButton onClick={handleDownloadMenuOpen}>
            <GetAppIcon />
          </IconButton>
        </Tooltip>
      );
    }, [handleDownloadMenuOpen, selected]);

    const analysisDate = useMemo(() => {
      if (analysisData instanceof BaselineLayerResult) {
        return analysisData.analysisDate;
      }
      if (analysisData instanceof PolygonAnalysisResult) {
        return analysisData.endDate;
      }
      return null;
    }, [analysisData]);

    const fileName = useMemo(() => {
      if (
        // Explicit condition for type narrowing
        !analysisData ||
        analysisData instanceof ExposedPopulationResult
      ) {
        return analysisData?.getTitle(t);
      }
      return generateAnalysisFilename(analysisData, analysisDate ?? null);
    }, [analysisData, analysisDate, t]);

    const handleDownloadCsv = useCallback((): void => {
      if (
        // Explicit condition for type narrowing
        !analysisData ||
        analysisData instanceof ExposedPopulationResult
      ) {
        return;
      }
      downloadCSVFromTableData(
        analysisData,
        translatedColumns,
        analysisDate ?? null,
        analysisResultSortByKey,
        analysisResultSortOrder,
      );
    }, [
      analysisData,
      analysisDate,
      analysisResultSortByKey,
      analysisResultSortOrder,
      translatedColumns,
    ]);

    const handleDownloadGeoJson = useCallback(() => {
      if (!analysisData) {
        return;
      }
      downloadToFile(
        {
          content: JSON.stringify(featureCollection),
          isUrl: false,
        },
        fileName ?? 'prism_extract',
        'application/json',
      );
    }, [analysisData, featureCollection, fileName]);

    const renderedDownloadAsCSVMenuItem = useMemo(() => {
      if (!doesLayerAcceptCSVDownload) {
        return null;
      }
      return (
        <MenuItem key="download-as-csv" onClick={handleDownloadCsv}>
          {t('Download as CSV')}
        </MenuItem>
      );
    }, [doesLayerAcceptCSVDownload, handleDownloadCsv, t]);

    return (
      <>
        {renderedDownloadButton}
        <Menu
          id="download-menu"
          anchorEl={downloadMenuAnchorEl}
          keepMounted
          open={Boolean(downloadMenuAnchorEl)}
          onClose={handleDownloadMenuClose}
        >
          {[
            renderedDownloadAsCSVMenuItem,
            <MenuItem key="download-as-geojson" onClick={handleDownloadGeoJson}>
              {t('Download as GeoJSON')}
            </MenuItem>,
          ]}
        </Menu>
      </>
    );
  },
);

interface AnalysisLayerSwitchItemDownloadOptionsProps {
  selected: boolean;
  analysisData?:
    | BaselineLayerResult
    | PolygonAnalysisResult
    | ExposedPopulationResult;
  analysisResultSortByKey: string | number;
  analysisResultSortOrder: 'asc' | 'desc';
}

export default AnalysisLayerSwitchItemDownloadOptions;
