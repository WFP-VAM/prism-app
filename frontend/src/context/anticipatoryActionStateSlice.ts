import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { LayerDefinitions } from 'config/utils';
import { AnticipatoryActionLayerProps } from 'config/types';
import type { CreateAsyncThunkTypes, RootState } from './store';

export const allWindowsKey = 'All';

// na/ny are not actually found in CSV, but defined not to cause confusion when calling the functions
export const AAcategory = ['na', 'ny', 'Leve', 'Moderado', 'Severo'] as const;
export type AACategoryType = typeof AAcategory[number];

export const AAPhase = ['na', 'ny', 'Ready', 'Set'] as const;
export type AAPhaseType = typeof AAPhase[number];

export interface AnticipatoryActionData {
  category: AACategoryType;
  district: string;
  index: string;
  month: string;
  phase: AAPhaseType;
  probability: string;
  trigger: string;
  triggerNB: string;
  triggerType: string;
  type: string;
  windows: string;
  yearOfIssue: string;
  date: string;
}

export const AAlayerKey = 'anticipatory_action';

const AACSVKeys: [string, string][] = [
  ['Category', 'category'],
  ['District', 'district'],
  ['Index', 'index'],
  ['Month', 'month'],
  ['Phase', 'phase'],
  ['Probability', 'probability'],
  ['Trigger', 'trigger'],
  ['Trigger_nb', 'triggerNB'],
  ['Trigger_type', 'triggerType'],
  ['Type', 'type'],
  ['Windows', 'windows'],
  ['Year_of_issue', 'yearOfIssue'],
];

function transform(data: any[], keys: [string, string][]) {
  const parsed = data
    .filter(x => !!x[keys[0][0]]) // filter empty rows
    .map(obj => {
      const entries = keys.map(k => [k[1], obj[k[0]]]);
      const month = obj.Month.padStart(2, '0');
      const year = obj.Year_of_issue.split('-')[0];

      // TODO: update this once CSV date format is revised
      const actualYear = Number(month) < 5 ? String(Number(year) + 1) : year;
      const date = `${actualYear}-${month}-01`;
      return Object.fromEntries([...entries, ['date', date]]);
    }) as AnticipatoryActionData[];

  const windows = [...new Set(parsed.map(x => x.windows))];
  // eslint-disable-next-line fp/no-mutating-methods
  const availableDates = [
    ...new Set(parsed.map(x => new Date(x.date).getTime())),
  ].sort();

  const groupedByDistrict = new Map<string, AnticipatoryActionData[]>();

  parsed.forEach(x => {
    const { district } = x;
    const rows = groupedByDistrict.get(district);
    groupedByDistrict.set(district, rows ? [...rows, x] : [x]);
  });

  const sortFn = (a: AnticipatoryActionData, b: AnticipatoryActionData) => {
    const aDate = new Date(a.date).valueOf();
    const bDate = new Date(b.date).valueOf();
    if (aDate > bDate) {
      return -1;
    }
    if (bDate > aDate) {
      return 1;
    }
    const aCatIndex = AAcategory.findIndex(x => x === a.category);
    const bCatIndex = AAcategory.findIndex(x => x === b.category);
    if (aCatIndex > bCatIndex) {
      return -1;
    }
    if (bCatIndex > aCatIndex) {
      return 1;
    }
    return 0;
  };

  const result = Object.fromEntries(
    // eslint-disable-next-line fp/no-mutating-methods
    Array.from(groupedByDistrict.entries()).map(x => [x[0], x[1].sort(sortFn)]),
  );

  const monitoredDistricts = Object.keys(result);

  return { data: result, windows, availableDates, monitoredDistricts };
}

type AnticipatoryActionState = {
  data: {
    [k: string]: AnticipatoryActionData[];
  };
  availableDates?: number[];
  monitoredDistricts: string[];
  windows: string[];
  selectedWindow: string;
  categoryFilters: Record<AACategoryType, boolean>;
  loading: boolean;
  error: string | null;
};

const initialState: AnticipatoryActionState = {
  data: {},
  windows: [],
  availableDates: undefined,
  monitoredDistricts: [],
  selectedWindow: allWindowsKey,
  categoryFilters: {
    Severo: true,
    Moderado: true,
    Leve: true,
    na: false,
    ny: false,
  },
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    data: {
      [k: string]: AnticipatoryActionData[];
    };
    windows: string[];
    availableDates: number[];
    monitoredDistricts: string[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const layer = LayerDefinitions[AAlayerKey] as AnticipatoryActionLayerProps;

  const url = layer.baseUrl;
  return new Promise<any>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(transform(results.data, AACSVKeys)),
      error: error => reject(error),
    }),
  );
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setSelectedWindow: (state, { payload }: PayloadAction<string>) => ({
      ...state,
      selectedWindow: payload,
    }),
    setCategoryFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<Record<'Leve' | 'Moderado' | 'Severo', boolean>>
      >,
    ) => ({
      ...state,
      categoryFilters: { ...state.categoryFilters, ...payload },
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: payload.data,
      windows: payload.windows,
      availableDates: payload.availableDates,
      monitoredDistricts: payload.monitoredDistricts,
    }));

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AnticipatoryActionDataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AnticipatoryActionWindowsSelector = (state: RootState) =>
  state.anticipatoryActionState.windows;

export const AnticipatoryActionAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AASelectedWindowSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedWindow;

export const AACategoryFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.categoryFilters;

// export actions
export const {
  setSelectedWindow,
  setCategoryFilters,
} = anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
