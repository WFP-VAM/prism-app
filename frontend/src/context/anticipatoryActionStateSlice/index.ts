import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { DateItem } from 'config/types';
import { appConfig } from 'config';
import { AAWindowKeys } from 'config/utils';
import type { CreateAsyncThunkTypes, RootState } from '../store';
import {
  AACategoryType,
  AAView,
  AnticipatoryActionData,
  AnticipatoryActionState,
  Vulnerability,
  allWindowsKey,
} from './types';
import {
  calculateMapRenderedDistricts,
  emptyWindows,
  parseAndTransformAA,
} from './utils';

const initialState: AnticipatoryActionState = {
  data: emptyWindows,
  availableDates: undefined,
  monitoredDistricts: [],
  filters: {
    selectedDate: undefined,
    selectedWindow: allWindowsKey,
    selectedIndex: '',
    categories: {
      Severe: true,
      Moderate: true,
      Mild: true,
      Normal: true,
      na: true,
      ny: true,
    },
  },
  markers: [],
  selectedDistrict: '',
  renderedDistricts: emptyWindows,
  windowRanges: { 'Window 1': undefined, 'Window 2': undefined },
  view: AAView.Home,
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    windowData: {
      data: AnticipatoryActionData;
      availableDates: DateItem[];
      windowKey: (typeof AAWindowKeys)[number];
      range: { start: string; end: string };
    }[];
    monitoredDistricts: { name: string; vulnerability: Vulnerability }[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const url = appConfig.anticipatoryActionUrl;

  return new Promise<any>((resolve, reject) => {
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(parseAndTransformAA(results.data)),
      error: error => reject(error),
    });
  });
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setAAFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<{
          selectedDate: string | undefined;
          selectedWindow: (typeof AAWindowKeys)[number] | typeof allWindowsKey;
          selectedIndex: string;
          categories: Partial<Record<AACategoryType, boolean>>;
        }>
      >,
    ) => {
      const { categories, ...rest } = payload;
      const newCategories =
        categories !== undefined
          ? { ...state.filters.categories, ...categories }
          : state.filters.categories;
      const newFilters = {
        ...state.filters,
        ...rest,
        categories: newCategories,
      };
      return {
        ...state,
        filters: newFilters,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: newFilters,
          data: state.data,
          windowRanges: state.windowRanges,
        }),
      };
    },
    setAASelectedDistrict: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['selectedDistrict']>,
    ) => ({
      ...state,
      selectedDistrict: payload,
    }),
    setAAMarkers: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['markers']>,
    ) => ({
      ...state,
      markers: payload,
    }),
    setAAView: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['view']>,
    ) => ({
      ...state,
      view: payload,
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => {
      const newData = Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.data]),
      ) as Record<(typeof AAWindowKeys)[number], AnticipatoryActionData>;
      const newRanges = Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.range]),
      ) as Record<
        (typeof AAWindowKeys)[number],
        { start: string; end: string }
      >;
      return {
        ...state,
        loading: false,
        data: newData,
        availableDates: Object.fromEntries(
          payload.windowData.map(x => [x.windowKey, x.availableDates]),
        ) as Record<(typeof AAWindowKeys)[number], DateItem[]>,
        windowRanges: newRanges,
        monitoredDistricts: payload.monitoredDistricts,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: state.filters,
          data: newData,
          windowRanges: newRanges,
        }),
      };
    });

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error: _error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AADataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AAAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AAMonitoredDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.monitoredDistricts;

export const AAFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.filters;

export const AARenderedDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.renderedDistricts;

export const AASelectedDistrictSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedDistrict;

export const AAMarkersSelector = (state: RootState) =>
  state.anticipatoryActionState.markers;

export const AAViewSelector = (state: RootState) =>
  state.anticipatoryActionState.view;

export const AAWindowRangesSelector = (state: RootState) =>
  state.anticipatoryActionState.windowRanges;

// export actions
export const { setAAFilters, setAASelectedDistrict, setAAMarkers, setAAView } =
  anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
