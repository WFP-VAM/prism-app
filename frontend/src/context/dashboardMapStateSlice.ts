import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Map as MaplibreMap, LngLatBounds } from 'maplibre-gl';

// Maplibre's map type contains some kind of cyclic dependency that causes an infinite loop in immers's change
// tracking. To save it off, we wrap it in a JS closure so that Redux just checks the function for changes, rather
// than recursively walking the whole object.
type MapGetter = () => MaplibreMap | undefined;

export interface DashboardMapState {
  maplibreMap: MapGetter;
  layers: any[];
  bounds: LngLatBounds | null;
  zoom: number | null;
}

const initialState: DashboardMapState = {
  maplibreMap: (() => {}) as MapGetter,
  layers: [],
  bounds: null,
  zoom: null,
};

export const dashboardMapStateSlice = createSlice({
  name: 'dashboardMapState',
  initialState,
  reducers: {
    setDashboardMap: (state, action: PayloadAction<MapGetter>) => ({
      ...state,
      maplibreMap: action.payload,
    }),
    addDashboardLayer: (state, action: PayloadAction<any>) => ({
      ...state,
      layers: [...state.layers, action.payload],
    }),
    removeDashboardLayer: (state, action: PayloadAction<string>) => ({
      ...state,
      layers: state.layers.filter(layer => layer.id !== action.payload),
    }),
    setDashboardLayers: (state, action: PayloadAction<any[]>) => ({
      ...state,
      layers: action.payload,
    }),
    setDashboardBounds: (state, action: PayloadAction<LngLatBounds>) => ({
      ...state,
      bounds: action.payload,
    }),
    setDashboardLocation: (
      state,
      action: PayloadAction<{ bounds: LngLatBounds; zoom: number }>,
    ) => ({
      ...state,
      bounds: action.payload.bounds,
      zoom: action.payload.zoom,
    }),
  },
});

// Getters
export const dashboardMapSelector = (state: {
  dashboardMapState: DashboardMapState;
}): MaplibreMap | undefined => state.dashboardMapState.maplibreMap();

export const dashboardLayersSelector = (state: {
  dashboardMapState: DashboardMapState;
}): any[] => state.dashboardMapState.layers;

export const dashboardBoundsSelector = (state: {
  dashboardMapState: DashboardMapState;
}): LngLatBounds | null => state.dashboardMapState.bounds;

export const dashboardZoomSelector = (state: {
  dashboardMapState: DashboardMapState;
}): number | null => state.dashboardMapState.zoom;

// Setters
export const {
  setDashboardMap,
  addDashboardLayer,
  removeDashboardLayer,
  setDashboardLayers,
  setDashboardBounds,
  setDashboardLocation,
} = dashboardMapStateSlice.actions;

export default dashboardMapStateSlice.reducer;
