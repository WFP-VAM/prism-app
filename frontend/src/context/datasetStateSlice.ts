import moment from 'moment';
import { orderBy } from 'lodash';
import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Dispatch } from 'redux';
import { ChartType, DatasetField } from 'config/types';
import { DEFAULT_DATE_FORMAT } from 'utils/name-utils';
import {
  EWSSensorData,
  EWSTriggersConfig,
  fetchEWSDataPointsByLocation,
} from 'utils/ews-utils';
import { fetchWithTimeout } from 'utils/fetch-with-timeout';
import type { CreateAsyncThunkTypes, RootState } from './store';
import { TableData } from './tableStateSlice';

export type EWSParams = {
  externalId: string;
  triggerLevels: EWSTriggerLevels;
  baseUrl: string;
};

type EWSTriggerLevels = {
  watchLevel: number;
  warning: number;
  severeWarning: number;
};

type EWSDataPointsRequestParams = EWSParams & {
  date: number;
};

type DatasetState = {
  data?: TableData;
  isLoading: boolean;
  datasetParams?: AdminBoundaryParams | EWSParams;
  chartType: ChartType;
  title: string;
};

const initialState: DatasetState = {
  isLoading: false,
  chartType: ChartType.Line,
  title: '',
};

type BoundaryProps = {
  code: number;
  level: string;
  name: string;
};

type BoundaryPropsDict = { [key: string]: BoundaryProps };

export type AdminBoundaryParams = {
  boundaryProps: BoundaryPropsDict;
  url: string;
  serverLayerName: string;
  id: string;
  datasetFields: DatasetField[];
};

export type AdminBoundaryRequestParams = AdminBoundaryParams & {
  startDate?: number;
  endDate?: number;
  level: string;
  adminCode: number;
};

export type DatasetRequestParams =
  | AdminBoundaryRequestParams
  | EWSDataPointsRequestParams;

type DataItem = {
  date: number;
  values: { [key: string]: string };
};

export enum TableDataFormat {
  DATE = 'date',
  TIME = 'time',
}

export const CHART_DATA_PREFIXES = { col: 'd', date: 'Date' };

const createTableData = (
  results: DataItem[],
  format: TableDataFormat,
): TableData => {
  const momentFormat =
    format === TableDataFormat.DATE ? DEFAULT_DATE_FORMAT : 'HH:mm';

  const sortedRows = orderBy(results, item => item.date).map(row => {
    const valuesObj = Object.values(row.values).reduce(
      (acc, item, index) => ({
        ...acc,
        [`${CHART_DATA_PREFIXES.col}${index + 1}`]: item,
      }),
      {},
    );

    return {
      [CHART_DATA_PREFIXES.date]: moment(row.date).format(momentFormat),
      ...valuesObj,
    };
  });

  const columns = Object.keys(sortedRows[0]);
  const initRow = Object.keys(results[0].values).reduce(
    (acc, item, index) => ({
      ...acc,
      [`${CHART_DATA_PREFIXES.col}${index + 1}`]: item,
    }),
    { [CHART_DATA_PREFIXES.date]: CHART_DATA_PREFIXES.date },
  );

  return {
    rows: [initRow, ...sortedRows],
    columns,
  };
};

export const loadEWSDataset = async (
  params: EWSDataPointsRequestParams,
  dispatch: Dispatch,
): Promise<TableData> => {
  const { date, externalId, triggerLevels, baseUrl } = params;

  const dataPoints: EWSSensorData[] = await fetchEWSDataPointsByLocation(
    baseUrl,
    date,
    dispatch,
    externalId,
  );

  const results: DataItem[] = dataPoints.map(item => {
    const [measureDate, value] = item.value;

    return {
      date: moment(measureDate).valueOf(),
      values: { measure: value.toString() },
    };
  });

  const tableData = createTableData(results, TableDataFormat.TIME);

  const EWSConfig = Object.entries(triggerLevels).reduce(
    (acc, [key, value]) => {
      const obj = {
        ...EWSTriggersConfig[key],
        values: tableData.rows.map(() => value),
      };

      return { ...acc, [key]: obj };
    },
    {},
  );

  const tableDataWithEWSConfig: TableData = {
    ...tableData,
    EWSConfig,
  };

  return new Promise<TableData>(resolve => resolve(tableDataWithEWSConfig));
};

type HDCResponse = {
  data: { [key: string]: number[] };
  date: string[];
  valids: number[];
};

/**
 * Executes a request to the WFP Humanitarian Data Cube (HDC)
 *
 * @return Promise with parsed object from request as DataItem array.
 */
const fetchHDC = async (
  url: string,
  datasetFields: DatasetField[],
  params: { [key: string]: any },
  dispatch: Dispatch,
): Promise<DataItem[]> => {
  const requestParamsStr = Object.entries(params)
    .map(([key, value]) => `${key}=${value}`)
    .join('&');

  // TODO - better error handling.
  let responseJson: HDCResponse = {
    data: {
      rfh: [100],
    },
    valids: [6.0],
    date: ['2022-03-21'],
  };
  const response = await fetchWithTimeout(
    `${url}?${requestParamsStr}`,
    dispatch,
    {},
    `Request failed to get HDC data at ${url}?${requestParamsStr}`,
  );

  // eslint-disable-next-line fp/no-mutation
  responseJson = await response.json();

  const dates: number[] = responseJson?.date?.map((date: string) =>
    moment(date).valueOf(),
  );

  return dates?.map((date, index) => {
    const values = datasetFields.reduce(
      (acc, field) => ({
        ...acc,
        [field.label]: responseJson.data[field.key]
          ? responseJson.data[field.key][index]
          : field.fallback,
      }),
      {},
    );

    return { date, values };
  });
};

// HDC API expects a parameter which depends on the layer
// rainfall = rfh; ndvi = vim; blended = rfb
const getVamParam = (serverLayerName: string): string => {
  if (serverLayerName.includes('vim') || serverLayerName.includes('viq')) {
    return 'vim';
  }
  if (serverLayerName.includes('blended')) {
    return 'rfb';
  }
  return 'rfh';
};

export const loadAdminBoundaryDataset = async (
  params: AdminBoundaryRequestParams,
  dispatch: Dispatch,
): Promise<TableData | undefined> => {
  const endDate = moment(params.endDate);
  const startDate = moment(params.startDate);

  const {
    url: hdcUrl,
    level,
    adminCode,
    serverLayerName,
    datasetFields,
  } = params;

  const endDateStr = endDate.format(DEFAULT_DATE_FORMAT);
  const startDateStr = startDate.format(DEFAULT_DATE_FORMAT);

  const hdcRequestParams = {
    level,
    admin_id: adminCode,
    coverage: 'full',
    vam: getVamParam(serverLayerName),
    start: startDateStr,
    end: endDateStr,
  };

  const results = await fetchHDC(
    hdcUrl,
    datasetFields,
    hdcRequestParams,
    dispatch,
  );

  const tableData = createTableData(results, TableDataFormat.DATE);
  return new Promise<TableData>(resolve => resolve(tableData));
};

export const loadDataset = createAsyncThunk<
  TableData | undefined,
  DatasetRequestParams,
  CreateAsyncThunkTypes
>(
  'datasetState/loadDataset',
  async (params: DatasetRequestParams, { dispatch }) => {
    return (params as AdminBoundaryRequestParams).id
      ? loadAdminBoundaryDataset(params as AdminBoundaryRequestParams, dispatch)
      : loadEWSDataset(params as EWSDataPointsRequestParams, dispatch);
  },
);

export const datasetResultStateSlice = createSlice({
  name: 'DatasetResultSlice',
  initialState,
  reducers: {
    clearDataset: (): DatasetState => initialState,
    setBoundaryParams: (
      state,
      { payload }: PayloadAction<AdminBoundaryParams>,
    ): DatasetState => ({
      ...state,
      datasetParams: payload,
    }),
    setDatasetTitle: (
      state,
      { payload }: PayloadAction<string>,
    ): DatasetState => ({ ...state, title: payload }),
    setDatasetChartType: (
      state,
      { payload }: PayloadAction<ChartType>,
    ): DatasetState => ({ ...state, chartType: payload }),
    updateAdminId: (
      state,
      { payload }: PayloadAction<string>,
    ): DatasetState => {
      if (!state.datasetParams) {
        return state;
      }

      const adminBoundaryParams = { ...state.datasetParams, id: payload };

      return { ...state, datasetParams: adminBoundaryParams };
    },
    setEWSParams: (
      state,
      { payload }: PayloadAction<EWSParams & { chartTitle: string }>,
    ): DatasetState => {
      const { externalId, chartTitle, triggerLevels, baseUrl } = payload;

      return {
        ...state,
        datasetParams: { externalId, triggerLevels, baseUrl },
        title: chartTitle,
      };
    },
  },
  extraReducers: builder => {
    builder.addCase(
      loadDataset.fulfilled,
      (
        { ...rest },
        { payload }: PayloadAction<TableData | undefined>,
      ): DatasetState => ({
        ...rest,
        data: payload,
        isLoading: false,
      }),
    );
    builder.addCase(loadDataset.rejected, state => ({
      ...state,
      loading: false,
    }));
    builder.addCase(loadDataset.pending, state => ({
      ...state,
      isLoading: true,
    }));
  },
});

export const datasetSelector = (state: RootState): DatasetState =>
  state.datasetState;
export const loadingDatasetSelector = (state: RootState): boolean =>
  state.datasetState.isLoading;

// Setters
export const {
  clearDataset,
  setBoundaryParams,
  updateAdminId,
  setDatasetTitle,
  setDatasetChartType,
  setEWSParams,
} = datasetResultStateSlice.actions;

export default datasetResultStateSlice.reducer;
