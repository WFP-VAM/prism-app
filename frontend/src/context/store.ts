import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateReduce from './analysisResultStateSlice';
import mapSelectionLayerStateReduce from './mapSelectionLayerStateSlice';
import mapBoundaryInfoStateReduce from './mapBoundaryInfoStateSlice';
import datasetResultStateReduce from './datasetStateSlice';
import mapTileLoadingStateReduce from './mapTileLoadingStateSlice';
import leftPanelStateReduce from './leftPanelStateSlice';
import opacityStateReduce from './opacityStateSlice';
import anticipatoryActionDroughtStateReduce from './anticipatoryAction/AADroughtStateSlice';
import anticipatoryActionStormStateReduce from './anticipatoryAction/AAStormStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateReduce,
  notificationState: notificationStateReduce,
  mapSelectionLayerStateSlice: mapSelectionLayerStateReduce,
  mapBoundaryInfoState: mapBoundaryInfoStateReduce,
  datasetState: datasetResultStateReduce,
  mapTileLoadingState: mapTileLoadingStateReduce,
  leftPanelState: leftPanelStateReduce,
  opacityState: opacityStateReduce,
  anticipatoryActionDroughtState: anticipatoryActionDroughtStateReduce,
  anticipatoryActionStormState: anticipatoryActionStormStateReduce,
});

export const store = configureStore({
  reducer,
  middleware: getDefaultMiddleware({
    // TODO: Instead of snoozing this check, we might want to
    // serialize the state
    serializableCheck: false,
    immutableCheck: {
      // do not check the following state branches for accidental
      // mutations. This saves a lot of time in dev mode (but has no
      // impact on production builds).
      ignoredPaths: [
        'mapState.boundaryRelations',
        'mapState.layersData',
        'analysisResultState.result',
        'serverState.availableDates',
      ],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
