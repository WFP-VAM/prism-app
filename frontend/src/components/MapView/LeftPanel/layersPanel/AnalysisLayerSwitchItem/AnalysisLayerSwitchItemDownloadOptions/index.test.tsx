import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';
import { store } from 'context/store';
import AnalysisLayerSwitchItemDownloadOptions from '.';

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <AnalysisLayerSwitchItemDownloadOptions
        analysisResultSortByKey="some sort by key"
        analysisResultSortOrder="asc"
        selected
      />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
