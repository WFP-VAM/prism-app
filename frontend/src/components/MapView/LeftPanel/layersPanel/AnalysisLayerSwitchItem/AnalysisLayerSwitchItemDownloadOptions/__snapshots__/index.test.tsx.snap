// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-tooltip
    title="Download"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </mock-tooltip>
  <mock-menu
    id="download-menu"
    keepmounted="true"
    open="false"
  >
    <mock-menuitem>
      Download as CSV
    </mock-menuitem>
    <mock-menuitem>
      Download as GeoJSON
    </mock-menuitem>
  </mock-menu>
</div>
`;
